### PR TITLE
Added missing pthread header functions and types

### LINKED ISSUE #505: Incomplete pthread header.
The current Smack pthread header seems incomplete, and it is not clear why.

If I look at the pthread functions declared by Smack’s header:
```
$ cat share/smack/include/pthread.h | grep "^\w\+\s\+p\w\+("      
pthread_t pthread_self(void);
int pthread_equal(pthread_t t1, pthread_t t2);
int pthread_join(pthread_t __th, void **__thread_return);
void pthread_exit(void *retval);
int pthread_mutexattr_init(pthread_mutexattr_t *attr);
int pthread_mutexattr_settype(pthread_mutexattr_t *attr, int type);
int pthread_mutex_init(pthread_mutex_t *mutex, const pthread_mutexattr_t *attr);
int pthread_mutex_lock(pthread_mutex_t *__mutex);
int pthread_mutex_unlock(pthread_mutex_t *__mutex);
int pthread_mutex_destroy(pthread_mutex_t *__mutex);
int pthread_cond_init(pthread_cond_t *__cond, pthread_condattr_t *__condattr);
int pthread_cond_wait(pthread_cond_t *__cond, pthread_mutex_t *__mutex);
int pthread_cond_signal(pthread_cond_t *__cond);
int pthread_cond_broadcast(pthread_cond_t *__cond);
int pthread_cond_destroy(pthread_cond_t *__cond);
int pthread_create(pthread_t *__newthread, __const pthread_attr_t *__attr,
```
I get far fewer than those listed, e.g., in the man page:
```
$ man pthread | col -b | grep "^\s\+\w\+\(\s\*\)\?\s\+\w\+(" 
     int pthread_create(pthread_t *thread, const pthread_attr_t *attr,
     int pthread_cancel(pthread_t thread)
     int pthread_detach(pthread_t thread)
     int pthread_equal(pthread_t t1, pthread_t t2)
     void pthread_exit(void *value_ptr)
     int pthread_join(pthread_t thread, void **value_ptr)
     int pthread_kill(pthread_t thread, int sig)
     int pthread_once(pthread_once_t *once_control, void
     pthread_t pthread_self(void)
     int pthread_setcancelstate(int state, int *oldstate)
     int pthread_setcanceltype(int type, int *oldtype)
     void pthread_testcancel(void)
     int pthread_attr_destroy(pthread_attr_t *attr)
     int pthread_attr_getinheritsched(const pthread_attr_t *attr,
     int pthread_attr_getschedparam(const pthread_attr_t *attr,
     int pthread_attr_getschedpolicy(const pthread_attr_t *attr, int *policy)
     int pthread_attr_getscope(const pthread_attr_t *attr, int
     int pthread_attr_getstacksize(const pthread_attr_t *attr, size_t
     int pthread_attr_getstackaddr(const pthread_attr_t *attr, void
     int pthread_attr_getdetachstate(const pthread_attr_t *attr, int
     int pthread_attr_init(pthread_attr_t *attr)
     int pthread_attr_setinheritsched(pthread_attr_t *attr, int inheritsched)
     int pthread_attr_setschedparam(pthread_attr_t *attr,
     int pthread_attr_setschedpolicy(pthread_attr_t *attr, int policy)
     int pthread_attr_setscope(pthread_attr_t *attr, int contentionscope)
     int pthread_attr_setstacksize(pthread_attr_t *attr, size_t stacksize)
     int pthread_attr_setstackaddr(pthread_attr_t *attr, void *stackaddr)
     int pthread_attr_setdetachstate(pthread_attr_t *attr, int detachstate)
     int pthread_mutexattr_destroy(pthread_mutexattr_t *attr)
     int pthread_mutexattr_getprioceiling(pthread_mutexattr_t *attr, int
     int pthread_mutexattr_getprotocol(pthread_mutexattr_t *attr, int
     int pthread_mutexattr_gettype(pthread_mutexattr_t *attr, int *type)
     int pthread_mutexattr_init(pthread_mutexattr_t *attr)
     int pthread_mutexattr_setprioceiling(pthread_mutexattr_t *attr, int
     int pthread_mutexattr_setprotocol(pthread_mutexattr_t *attr, int
     int pthread_mutexattr_settype(pthread_mutexattr_t *attr, int type)
     int pthread_mutex_destroy(pthread_mutex_t *mutex)
     int pthread_mutex_init(pthread_mutex_t *mutex,
     int pthread_mutex_lock(pthread_mutex_t *mutex)
     int pthread_mutex_trylock(pthread_mutex_t *mutex)
     int pthread_mutex_unlock(pthread_mutex_t *mutex)
     int pthread_condattr_destroy(pthread_condattr_t *attr)
     int pthread_condattr_init(pthread_condattr_t *attr)
     int pthread_cond_broadcast(pthread_cond_t *cond)
     int pthread_cond_destroy(pthread_cond_t *cond)
     int pthread_cond_init(pthread_cond_t *cond, const pthread_condattr_t
     int pthread_cond_signal(pthread_cond_t *cond)
     int pthread_cond_timedwait(pthread_cond_t *cond, pthread_mutex_t *mutex,
     int pthread_cond_wait(pthread_cond_t *, pthread_mutex_t *mutex)
     int pthread_rwlock_destroy(pthread_rwlock_t *lock)
     int pthread_rwlock_init(pthread_rwlock_t *lock,
     int pthread_rwlock_rdlock(pthread_rwlock_t *lock)
     int pthread_rwlock_tryrdlock(pthread_rwlock_t *lock)
     int pthread_rwlock_trywrlock(pthread_rwlock_t *lock)
     int pthread_rwlock_unlock(pthread_rwlock_t *lock)
     int pthread_rwlock_wrlock(pthread_rwlock_t *lock)
     int pthread_rwlockattr_destroy(pthread_rwlockattr_t *attr)
     int pthread_rwlockattr_getpshared(const pthread_rwlockattr_t *attr,
     int pthread_rwlockattr_init(pthread_rwlockattr_t *attr)
     int pthread_rwlockattr_setpshared(pthread_rwlockattr_t *attr, int
     int pthread_key_create(pthread_key_t *key, void (*routine)(void *))
     int pthread_key_delete(pthread_key_t key)
     void * pthread_getspecific(pthread_key_t key)
     int pthread_setspecific(pthread_key_t key, const void *value_ptr)
     int pthread_atfork(void (*prepare)(void), void (*parent)(void),
     void pthread_cleanup_pop(int execute)
     void pthread_cleanup_push(void (*routine)(void *), void *routine_arg)
```
What gives?

I can understand not having great models for many of these functions, but why aren’t they even declared?

Without declarations, many codes will not even compile with Smack’s pthread header.

### PR DESCRIPTION
**Pull request by zvonimir**
I think this now completes it.

Closes #505

### CODE REVIEW

**Comment by michael-emmi** [comment]
Can you remind me why we are even including headers for libraries like pthead in the first place? Why can’t the system `pthread.h` header be used here?

**Comment by zvonimir** [comment]
Because we leverage pthread types to store the needed information for modeling. Hence, we define these types to suit our purpose, and our definitions might not match the systems ones.

File: share/smack/include/pthread.h [hunk 1]
```diff
@@ -18,73 +20,95 @@
...
   { UNLOCKED, INITIALIZED }
 
-// Declare corral primitive procedures
-//__SMACK_INIT(corral_primitives);
-
-// Initialize thread status tracking variables (OS level behavior)
-//__SMACK_INIT(thread);
-
+// Helper functions
 void __VERIFIER_atomic_begin();
```
Comment (michael-emmi) [inline share/smack/include/pthread.h]: While we’re cleaning this up, can we move `__VERIFIER_atomic_*` to a more-appropriate place, e.g., `smack.h`?

File: share/smack/include/pthread.h [hunk 2]
```diff
@@ -18,73 +20,95 @@
...
-
 void __VERIFIER_atomic_end();
+void __call_wrapper(pthread_t *__newthread, void *(*__start_routine)(void *),
+                    void *__arg);
 
-pthread_t pthread_self(void);
-
+int pthread_create(pthread_t *__newthread, __const pthread_attr_t *__attr,
```
Comment (michael-emmi) [inline share/smack/include/pthread.h]: Perhaps we should alphabetize all declarations here, i.e., start with the `pthread_attr` declarations?

Comment (zvonimir) [reply share/smack/include/pthread.h]: Done.

File: share/smack/include/pthread.h (hunk 1 above)
Comment (zvonimir) [reply share/smack/include/pthread.h]: Done.

**Pull request merged**